stainless_extraction/src/std_items.rs

```diff
@@ -25,184 +34,121 @@ pub(super) enum StdItem {
-
-    // Register additional items from the rust standard library
-    let std_crate_num = this.item_to_def(BeginPanicFn).krate;
-    this.register_items_from_crate(tcx, &[BeginPanicFmtFn], std_crate_num);
-
-    // Register items from stainless crate
-    let stainless_sym = Symbol::intern("stainless");
-    let stainless_crate_num = match tcx
+    let crate_nums: HashMap<String, CrateNum> = tcx
       .crates()
       .iter()
-      .find(|&&cnum| tcx.crate_name(cnum) == stainless_sym)
-      .copied()
-    {
-      Some(crate_num) => crate_num,
-      None => tcx.sess.fatal(
-        "Couldn't find stainless library. Make sure it is included using 'extern crate stainless;'",
-      ),
-    };
-    this.register_items_from_crate(tcx, STAINLESS_ITEMS, stainless_crate_num);
-
-    this
-  }
+      .filter_map(|&cnum| {
+        let name = tcx.crate_name(cnum).to_string();
+        CRATE_NAMES.contains(&name.as_str()).then(|| (name, cnum))
+      })
+      .collect();
 
-  fn make_def_id(cnum: CrateNum, index: usize) -> DefId {
-    DefId {
-      krate: cnum,
-      index: DefIndex::from_usize(index),
+    if !crate_nums.contains_key("stainless") {
+      tcx.sess.fatal(
+        "Couldn't find stainless library. Make sure it is included using 'extern crate stainless;'",
+      )
     }
-  }
 
-  fn register_items_from_lang_items(&mut self, tcx: TyCtxt, items: &[StdItem]) {
-    for item in items {
-      let def_id = tcx.require_lang_item(item.lang_item().unwrap(), None);
-      self.item_to_def[item.index()] = def_id;
-      self.def_to_item.insert(def_id, *item);
+    // Register rust lang items
+    for item in LangItem::iter() {
+      let def_id = tcx.require_lang_item(item.rust_lang_item(), None);
+      this.did_to_lang_item.insert(def_id, *item);
     }
-  }
 
-  // HACK(gsps): A horrible way to do this, but rustc -- as far as I can tell -- does not expose
-  // any API for finding a specific item in, or enumerating all items of a crate.
-  fn register_items_from_crate(&mut self, tcx: TyCtxt, items: &[StdItem], cnum: CrateNum) {
-    let mut items: HashMap<Symbol, StdItem> = items
-      .iter()
-      .map(|item| (Symbol::intern(item.name()), *item))
-      .collect();
-
-    for index in 0.. {
-      if items.is_empty() {
-        break; // We found DefIds for all items, we're done.
-      }
-      let def_id = Self::make_def_id(cnum, index);
-      if let Some(ref name_sym) = tcx
-        .def_path(def_id)
-        .data
-        .last()
-        .map(|data| data.data.as_symbol())
-      {
-        if let Some(item) = items.remove(name_sym) {
-          self.item_to_def[item.index()] = def_id;
-          self.def_to_item.insert(def_id, item);
-        }
-      }
-    }
+    this
   }
 
   #[inline]
-  pub(super) fn item_to_def<I: Into<StdItem>>(&self, item: I) -> DefId {
-    let item = item.into();
-    self.item_to_def[item.index()]
+  pub(super) fn def_to_item_opt(&self, tcx: &TyCtxt<'_>, def_id: DefId) -> Option<StdItem> {
```

Comment:
I'm not at all happy, that the `&tcx` has to be passed in here. I tried to store a reference to it in the `StdItems` struct but then I had lifetime problems:

```
error[E0515]: cannot return value referencing function parameter `tcx`
  --> stainless_extraction/src/lib.rs:79:3
   |
60 |   let std_items = Rc::new(StdItems::collect(&tcx));
   |                                             ---- `tcx` is borrowed here
...
79 |   symbols
   |   ^^^^^^^ returns a value referencing data owned by the current function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0515`.
error: could not compile `stainless_extraction`.

To learn more, run the command again with --verbose.
```

Reply:
Forgot that `TyCtx` is copy! 🎉